### PR TITLE
Improve flexibility in groundwater setup

### LINKED ISSUE #347: Support variable/fixed internal time stepping kinematic wave lateral subsurface flow
### Feature type

Changing existing functionality

### Improvement Description

The current implementation of kinematic wave lateral subsurface flow lacks is solved at the Model time step and thus lacks support for a variable (based on Courant number) or fixed internal time stepping compared to the river and overland flow kinematic wave components.

### Implementation Description

Add internal time stepping for kinematic wave lateral subsurface flow component.

### Additional Context

_No response_

### LINKED ISSUE #348: Enhanced Flexibility in Groundwater Setup
### Feature type

Changing existing functionality

### Improvement Description

The current groundwater setup lacks the necessary flexibility needed for various simulation scenarios. Specifically, the inability to switch between kinematic wave and Darcian flow without specifying additional groundwater boundaries limits the versatility of the model.

### Implementation Description

Revise Groundwater Setup for Flexibility: Overhaul the current groundwater setup to allow seamless switching between kinematic wave and Darcian flow regimes without the requirement of specifying additional groundwater boundaries. This enhancement will provide users with the flexibility to adapt the model to different hydrological conditions and simulation objectives without significant overhead.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by vers-w**
## Issue addressed
Fixes #348, #347

## Explanation
The groundwater concept `GroundwaterFlow` and lateral subsurface flow concept `LateralSSF` can now make use of the same head-dependent flux `boundaries` (`recharge`, `river` and `drain`).  As part of `LateralSSF` the variable `head` is now also computed, also when head-dependent flux `river` and `drain` boundaries are not included, as this can be useful output. Additionally, the variable naming between the `GroundwaterFlow` and `LateralSSF` concept is more consistent for clarity and more efficient code. The `ConfinedAquifer` concept has been removed, this functionality is not exposed as part of a wflow model and is not expected to be used anytime soon.

Supporting internal timestepping for `LateralSSF`, including an alpha coefficient to enhance the stability of the simulation, setting the alpha coefficient lower than 1.0 can be useful when head-dependent flux boundaries are included.

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `master`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed
- [x] Updated changelog.qmd if needed

### CODE REVIEW

File: Wflow/src/routing/subsurface/groundwater.jl [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+        dh, exfilt = water_table_change(soil, net_flux, specific_yield[i], i)
+        head[i] += dh
+        exfiltwater[i] += exfilt
+    end
+    # Set constant head (dirichlet) boundaries
+    gwf.variables.head[gwf.constanthead.index] .= gwf.constanthead.variables.head
+    # Make sure no heads ends up below an unconfined aquifer bottom
+    gwf.variables.head .= minimum_head(gwf)
+    # Adjust exfiltration rate for constant head boundaries
+    exfiltwater[gwf.constanthead.index] .= 0.0
```
Comment (JoostBuitink) [inline Wflow/src/routing/subsurface/groundwater.jl:556]: I believe this functionality is not new, but does it make sense to set the exfiltration to 0 on the cells with constant head boundaries?

File: Wflow/src/routing/timestepping.jl [hunk 2]
```diff
@@ -1,10 +1,10 @@
 
-"Timestepping for solving kinematic wave and local inertial river and overland flow routing."
+"Timestepping for river, overland and subsurface flow routing."
 @with_kw struct TimeStepping
     stable_timesteps::Vector{Float64} = Float64[]
     dt_fixed::Float64 = 0.0
     adaptive::Bool = true
-    cfl::Float64 = 0.70
+    alpha_coefficient::Float64 = 1.0
```
Comment (JoostBuitink) [inline Wflow/src/routing/timestepping.jl:7]: Is this change in default value properly documented?

Comment (JoostBuitink) [reply Wflow/src/routing/timestepping.jl:7]: I see that the default for land, overland and subsurface are still defined as before in the `config_structure.jl`; so feel free to ignore this comment.

File: Wflow/src/Wflow.jl [hunk 3]
```diff
@@ -196,16 +196,18 @@ include("surfacewater/runoff.jl")
...
-include("routing/routing_process.jl")
+include("routing/subsurface/connectivity.jl")
+include("routing/subsurface/groundwater.jl")
+include("routing/subsurface/lateral_subsurface_flow.jl")
+include("routing/subsurface/subsurface_process.jl")
+include("routing/subsurface/boundary_conditions.jl")
+include("routing/surface/reservoir.jl")
+include("routing/surface/surface_kinwave.jl")
+include("routing/surface/surface_local_inertial.jl")
+include("routing//surface/surface_routing.jl")
```
Comment (JoostBuitink) [inline Wflow/src/Wflow.jl]:
```suggestion
include("routing/surface/surface_routing.jl")
```

File: Wflow/test/groundwater.jl [hunk 4]
```diff
@@ -388,23 +335,24 @@ end
...
-        @test all(diff(gwf.aquifer.variables.head) .> 0.0)
+        @test all(diff(gwf.variables.head) .> 0.0)
     end
 
     head_analytical = [
         transient_aquifer_1d(x, time, conductivity, specific_yield, aquifer_length, beta) for x in xc
     ]
-    difference = gwf.aquifer.variables.head .- head_analytical
+    difference = gwf.variables.head .- head_analytical
     # @test all(difference .< ?)  #TODO
```
Comment (JoostBuitink) [inline Wflow/test/groundwater.jl]: I know this was already there, but perhaps nice to pick this up as well?

File: Wflow/test/run_sbm_gwf_piave.jl [hunk 5]
```diff
@@ -38,58 +38,58 @@
         @test domestic.demand.demand_net[[1, end]] ≈ [0.3802947998046875, 0.0]
         @test domestic.variables.returnflow[[1, end]] ≈ [0.2209725379943848, 0.0]
         @test reservoir.variables.waterlevel ≈
-              [23.968537463000757, 32.68607771649563, 39.97018425222191]
-        @test reservoir.variables.storage ≈ [1.5531612276024494e8, 4.28e7, 7.16e7]
+              [23.970590790933628, 32.68607771649563, 39.97018425222192]
+        @test reservoir.variables.storage ≈ [1.5532942832524955e8, 4.28e7, 7.16e7]
         @test reservoir.variables.outflow_av ≈
-              [3.248673046140208, 8.352196766583088, 29.02990124474297]
+              [3.2489121397532985, 8.556416216129914, 28.17032166161074]
```
Comment (JoostBuitink) [inline Wflow/test/run_sbm_gwf_piave.jl:44]: How is it possible that the reservoir outflow (mostly the last one) changes so much (from `29.03` to `28.17`?

File: Wflow/test/run_sbm_gwf_piave.jl [hunk 6]
```diff
@@ -38,58 +38,58 @@
...
         @test nonpaddy.variables.demand_gross[[32, 38, 41]] ≈
               [0.0, 4.264347104462701, 5.022735931644931]
         @test reservoir.variables.waterlevel ≈
-              [23.963488224416217, 32.68607771649562, 39.970184252221905]
-        @test reservoir.variables.storage ≈ [1.552834064233491e8, 4.28e7, 7.16e7]
+              [23.964702651630343, 32.686077716495625, 39.97018425222192]
+        @test reservoir.variables.storage ≈ [1.5529127318256468e8, 4.28e7, 7.16e7]
         @test reservoir.variables.outflow_av ≈
-              [3.248477758506481, 9.78606013238214, 36.7768455543205]
+              [3.2489840968665207, 9.467087432483991, 38.61905891069846]
```
Comment (JoostBuitink) [inline Wflow/test/run_sbm_gwf_piave.jl:77]: Same as the comment above, but now from `36.78` to `38.62`

File: Wflow/test/sbm_piave_nonirri-demand_config.toml [hunk 7]
```diff
@@ -49,6 +49,7 @@ atmosphere_air__temperature = "temp"
 
 [input.static]
 land_surface__slope = "Slope"
+land_surface__elevation = "wflow_dem"
```
Comment (JoostBuitink) [inline Wflow/test/sbm_piave_nonirri-demand_config.toml:52]: Just wondering now, does this also need hydromt_wflow changes? Also for the new `ssf`-state name

File: Wflow/src/routing/subsurface/subsurface_process.jl [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+end
+
+"""
+    kinematic_wave_ssf(q_in, q_prev, zi_prev, q_net_bnds, slope, sy, d, dt, dx, dw, q_max, kh_profile, soil, i)
+
+Kinematic wave for lateral subsurface flow for a single cell and timestep. The hydraulic
+conductivity profile `kh_profile` is either `KhExponential` or `KhExponentialConstant`.
+
+Returns lateral subsurface flow `q`, water table depth `zi`, exfiltration rate `exfilt`
+and dynamic specific yield `sy_d`.
```
Comment (JoostBuitink) [inline Wflow/src/routing/subsurface/subsurface_process.jl]: This docstring is outdated, it now returns `net_flux`

File: Wflow/test/run_sbm_piave.jl [hunk 9]
```diff
@@ -16,76 +16,76 @@
...
-        230.50784825603955,
-        203.15747076518738,
-        274.1950442509512,
-        161.21111379880762,
-        166.53224821211137,
-        149.289171819346,
-        124.52721206865995,
-        107.80542681736212,
-        169.00991650585908,
-        106.49205508983894,
```
Comment (JoostBuitink) [inline Wflow/test/run_sbm_piave.jl:40]: Do you know why these values changed? And also why they did not change for the moselle tests?

File: Wflow/test/run_sbm_piave.jl [hunk 10]
```diff
@@ -132,51 +132,51 @@ end
         @test domestic.demand.demand_net[[1, end]] ≈ [0.3802947998046875, 0.0]
         @test domestic.variables.returnflow[[1, end]] ≈ [0.2209725379943848, 0.0]
         @test reservoir.variables.waterlevel ≈
-              [29.259064145949047, 32.68607771649561, 39.97018425222191]
-        @test reservoir.variables.storage ≈ [1.8959873566574994e8, 4.28e7, 7.16e7]
+              [29.259144530899885, 32.68607771649562, 39.970184252221905]
+        @test reservoir.variables.storage ≈ [1.8959925656023118e8, 4.28e7, 7.16e7]
         @test reservoir.variables.outflow_av ≈
-              [4.839239564226234, 9.69159914425381, 57.18888358068246]
+              [4.839249448770597, 9.70890592273304, 57.64196655435958]
```
Comment (JoostBuitink) [inline Wflow/test/run_sbm_piave.jl:138]: Same as other comments: it doesnt sound intuitive that the reservoir storage remains the same, yet the outflow is different

**Review by JoostBuitink — CHANGES_REQUESTED** [review]

File: Wflow/test/run_sbm_gwf_piave.jl (hunk 5 above)
Comment (vers-w) [reply Wflow/test/run_sbm_gwf_piave.jl:44]: This is related to the `model` setting `subsurface_water_flow__alpha_coefficient` in the TOML file (internal timestepping).

File: Wflow/test/run_sbm_gwf_piave.jl (hunk 6 above)
Comment (vers-w) [reply Wflow/test/run_sbm_gwf_piave.jl:77]: See comment above (internal timestepping).

File: Wflow/test/sbm_piave_nonirri-demand_config.toml (hunk 7 above)
Comment (vers-w) [reply Wflow/test/sbm_piave_nonirri-demand_config.toml:52]: Yes, this also needs hydromt_wflow changes. I did add this to the wflow release notes in the changelog. After this PR has been approved I can report an issue requesting these changes.

File: Wflow/test/run_sbm_piave.jl (hunk 9 above)
Comment (vers-w) [reply Wflow/test/run_sbm_piave.jl:40]: Yes, because of the `kinematic_wave__adaptive_time_step_flag` setting in the TOML internal timestepping is also applied to kinematic wave lateral subsurface flow. For Piave this results in timestep < 1.0 day, for the Moselle this is not the case.

File: Wflow/test/run_sbm_piave.jl (hunk 10 above)
Comment (vers-w) [reply Wflow/test/run_sbm_piave.jl:138]: Yes, this is because of internal timestepping, see also comment above.

File: Wflow/src/routing/subsurface/groundwater.jl (hunk 1 above)
Comment (vers-w) [reply Wflow/src/routing/subsurface/groundwater.jl:556]: Yes, I think it makes sense. Note that these cells are part of the computations, `head` is also reset to the **constant** head boundary values while `exfiltwater` is the result of a change in `head`.

File: Wflow/test/groundwater.jl (hunk 4 above)
Comment (vers-w) [reply Wflow/test/groundwater.jl]: I have looked briefly into this and differences are quite large, maybe also why this has been marked as a TODO. For now I did remove this part the test and will report an issue so we can look into this further.

File: Wflow/test/run_sbm_gwf_piave.jl (hunk 5 above)
Comment (JoostBuitink) [reply Wflow/test/run_sbm_gwf_piave.jl:44]: Thanks for clarifying. Perhaps I miss something, but I don't yet understand how the subsurface flow timestepping affects the reservoir outflow (also given that the storage remains the same)?

**Review by JoostBuitink — APPROVED** [review]
Just one remaining question; mostly for my own understanding (hence why I do the approve now as well). Also left the comment related to hydromt_wflow open to ensure we don't forget!

File: Wflow/test/sbm_piave_nonirri-demand_config.toml (hunk 7 above)
Comment (vers-w) [reply Wflow/test/sbm_piave_nonirri-demand_config.toml:52]: https://github.com/Deltares/hydromt_wflow/issues/727

File: Wflow/test/run_sbm_gwf_piave.jl (hunk 5 above)
Comment (vers-w) [reply Wflow/test/run_sbm_gwf_piave.jl:44]: Running without the `subsurface_water_flow__alpha_coefficient` setting gives the previous results for reservoir `outflow_av`. Running with internal timestepping results in smaller `inflow` amounts for the reservoir. The reason that storage remains the same for two reservoirs is that these are `maxstorage` values.
